### PR TITLE
Migrate le/gt/ge/eq/ne from the TH to Aten. Added support of type promotion.

### PR DESCRIPTION
**Pull request by ifedan**
#24606 Migrate ne and ne_ from the TH to Aten (CUDA)
#24740 Migrate ne and ne_ from the TH to Aten (CPU)
#24573 Migrate gt and gt_ from the TH to Aten (CUDA)
#24709 Migrate gt and gt_ from the TH to Aten (CPU)
#24556 Migrate eq and eq_ from the TH to Aten (CUDA)
#24696 Migrate eq and eq_ from the TH to Aten (CPU)
#24568 Migrate ge and ge_ from the TH to Aten (CUDA)
#24703 Migrate ge and ge_ from the TH to Aten (CPU)
#24582 Migrate le and le_ from the TH to Aten (CUDA)
#24719 Migrate le and le_ from the TH to Aten (CPU)

Performance characteristics are similar to #25998

This PR migrates comparison ops from TH to ATen and adds type promotion in the same way as in #25998

### CODE REVIEW

**Comment by zou3519** [comment]
You should note in the commit message that this adds some type promotion behavior

**Comment by zou3519** [comment]
"All the comparison operation will follow type promotion described" this makes it vague about if we've added type promotion or not. It would be good to be explicit and say "This PR migrates comparison ops from TH to ATen and adds type promotion in the same way as in #25998". Would also be good to update the PR title to say something about type promotion.

File: aten/src/THC/THCTensorMathCompare.cuh [hunk 1]
```diff
@@ -7,67 +7,6 @@
...
-  TensorNEValueOp(T v) : value(v) {}
-  __device__ __forceinline__ void operator()(TOut* out, T* in) {
-    *out = ScalarConvert<bool, TOut>::to(THCNumerics<T>::ne(*in, value));
-  }
-
-  const T value;
-};
-
 template<typename ScalarTypeOut, typename ScalarType, typename TensorTypeOut, typename TensorType, class Op>
 void THC_logicalValue(THCState *state,
```
Comment (zou3519) [inline aten/src/THC/THCTensorMathCompare.cuh]: I'm not sure if we still need this. However, let's worry about that in the future and not in this PR.

File: test/test_torch.py [hunk 2]
```diff
@@ -6207,20 +6207,47 @@ def test_comparison_ops(self):
...
     def test_comparison_ops_must_take_bool_output(self):
         with self.assertRaisesRegex(RuntimeError, 'The output tensor of lt must be a bool'):
-            torch.lt(torch.tensor([True]), torch.tensor([False]), out=torch.empty(1, dtype=torch.uint8))
+            for op in [torch.lt, torch.le, torch.gt, torch.ge, torch.eq, torch.ne]:
+                op(torch.tensor([True]), torch.tensor([False]), out=torch.empty(1, dtype=torch.uint8))
 
     def test_inplace_comparison_ops_require_inputs_have_same_dtype(self):
         with self.assertRaisesRegex(RuntimeError, 'Expected object of scalar type'):
             torch.tensor([1], dtype=torch.int).lt_(torch.tensor([2], dtype=torch.long))
+            torch.tensor([1], dtype=torch.int).le_(torch.tensor([2], dtype=torch.long))
```
Comment (zou3519) [inline test/test_torch.py]: Please clean up these tests by not repeating yourself. If we make this a for loop, it only needs to be 3 lines instead of 6

**Review by zou3519 — APPROVED** [review]
We can repeat ourselves less in the tests but otherwise lgtm. Good work in deleting all those lines of code!

**Comment by ifedan** [comment]
> We can repeat ourselves less in the tests but otherwise lgtm. Good work in deleting all those lines of code!

Test changed based on comment

**Comment by mruberry** [comment]
This PR looks super cool but I think it was breaking cuDNN test builds. I had to unland it. I think you just need to modify test_overwrite_module_params_on_conversion_cpu_cuda.

**Pull request closed without merging**